test/expect-tests/memo/memoize_tests.ml

```diff
@@ -1420,3 +1419,100 @@ let%expect_test "error handling and duplicate exceptions" =
     Calling f 0
     Error [ "(Failure 42)" ]
     |}]
+
+let%expect_test "errors are cached" =
+  Printexc.record_backtrace false;
+  let f =
+    Memo.create_hidden "area of a square"
+      ~input:(module Int)
+      (fun x ->
+        printf "Started evaluating %d\n" x;
+        if x < 0 then failwith (sprintf "Negative input %d" x);
+        let res = x * x in
+        printf "Evaluated %d: %d\n" x res;
+        Memo.Build.return res)
+  in
+  evaluate_and_print f 5;
+  evaluate_and_print f (-5);
+  [%expect
+    {|
+    Started evaluating 5
+    Evaluated 5: 25
+    f 5 = Ok 25
+    Started evaluating -5
+    f -5 = Error [ { exn = "(Failure \"Negative input -5\")"; backtrace = "" } ]
+    |}];
+  evaluate_and_print f 5;
+  evaluate_and_print f (-5);
+  (* Note that we do not see any "Started evaluating" messages because both [Ok]
+     and [Error] results have been cached. *)
+  [%expect
+    {|
+    f 5 = Ok 25
+    f -5 = Error [ { exn = "(Failure \"Negative input -5\")"; backtrace = "" } ]
+    |}]
+
+let%expect_test "errors work with early cutoff" =
+  let divide =
+    Memo.create "divide 100 by input"
+      ~input:(module Int)
+      ~visibility:Hidden
+      ~output:(Allow_cutoff (module Int))
+      ~doc:""
+      (fun x ->
+        let+ (_ : Memo.Run.t) = Memo.current_run () in
+        printf "[divide] Started evaluating %d\n" x;
+        (if x > 100 then
+          (* We create a new exception each type on purpose, to break the cutoff *)
+          let exception Input_too_large in
```

Comment:
I'd use an exception where it's more explicit that it is fresh on each run. For instance `Input_too_large (Memo.current_run ())`.

Reply:
Sounds good, done!